HACBS - Application workflow visualization

karthikjeeyar opened the pull request:
## Fixes 
<!-- For e.g Fixes: https://issues.redhat.com/browse/HAC-XXX -->
https://issues.redhat.com/browse/HACBS-794

## Description
<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

This PR contains the Application Workflow visualization.

The node data model is mapped with the following CRs

- Source node  - > **Component**
- Build  -> Application **PipelineRuns** of type builds
- Component Integration -> **IntegrationTestScenarios** with component `context`
- Application Integration -> **IntegrationTestScenarios** with application `context`
- Static Environments -> **Environments**
- Managed Environments -> **ReleaseLinks**

## Type of change
<!-- Please delete options that are not relevant. -->

- [x] Feature


## Screen shots / Gifs for design review 
<!-- If change affects UI in any way, tag relevant UX people and add screenshots/gifs  -->
Disabled nodes:


![image](https://user-images.githubusercontent.com/9964343/182138348-9bdef191-be42-47bf-85c8-875e8a087b47.png)

Variable width nodes:


![image](https://user-images.githubusercontent.com/9964343/182140975-46ee392f-8af4-4405-88ef-9487bacaefc6.png)


## How to test or reproduce?
<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

<!-- - [ ] Test A -->
<!-- - [ ] Test B -->

1. Create Application, componetns, IntegrationTestScenorios, Envirnoments and releaseLinks CR.

use this [link](https://github.com/karthikjeeyar/hacbs-sample-resources/tree/main/examples) to refer the samples

##Unit Tests:
<!-- **Test Configuration(s)**: -->

```
   visualization-utils.spec.ts
  getWorkflowNodes
    ✓ expect to return empty array (2 ms)
    ✓ expect to return same number of nodes for linear workflows
    ✓ expect to contain a spacer node for parallel to parallel workflows (13 ms)
  getTopologyNodesEdges
    ✓ expect to return nodes and edges empty array (1 ms)
    ✓ expect to return nodes and edges for a workflow (2 ms)
  dagtoNodes
    ✓ should return empty for invalid values (1 ms)
    ✓ should return empty array if dag instance does not have vertices (1 ms)
    ✓ should return nodes if dag instance with vertices
  workflowToNodes
    ✓ should return empty array for workflow with no steps
    ✓ should return empty for invalid values (1 ms)
  sortEnvironments
    ✓ should return empty array for invalid values (1 ms)
    ✓ should return sorted array of environments (1 ms)
  getLastEnvironment
    ✓ should return default environment for invalid values (1 ms)
    ✓ should return the last environment in the list```
```


```
 useAppWorkflowData.spec.ts
  ✓ should early return empty values if any one of the resources is not loaded (12 ms)
  ✓ should return the basic workflow nodes (4 ms)
  ✓ should return disabled nodes if the resources are not created (2 ms)
  ✓ Abstract nodes should contain staic name (2 ms)
  ✓ Non Abstract nodes should contain dynamic name (4 ms)
  ✓ nodes should be disabled / enabled based on the availability of resources (2 ms)
  ✓ label should change based on the availability of resources (2 ms)
```

## Browser conformance: 
<!-- To mark tested browsers, use [x] -->
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge


<!-- ## Checklist: -->

<!-- 
- [ ] Code follows the style guidelines
- [ ] Self-reviewed the code
- [ ] Added comments in hard-to-understand areas
- [ ] Made corresponding changes to the documentation
- [ ] Changes generate no new warnings
- [ ] Added tests that prove this fix is effective or that the feature works
- [ ] New and existing unit tests pass locally with new changes
- [ ] Any dependent changes have been merged and published in downstream modules 
-->

karthikjeeyar commented:
cc: @Ranelim @MariaLeonova

karthikjeeyar commented:
/assign @christianvogt

karthikjeeyar commented:
/retest

christianvogt commented:
fyi @Ranelim @MariaLeonova

Ranelim commented:
@karthikjeeyar [karthikjeeyar](https://github.com/karthikjeeyar)
Thank you, it looks awesome!!

Some adjstments:
1. I believes the Node design PF is working on does not have a fixed for all the nodes. The node width changes according to the content within it. PF designers did say there is a default max number of characters, which might be 14 displayed characters within a node. A string that exceeds that number will display the beginning and the end of the string, with "..." in the middle. Not sure how many characters are at the end or in total by default.
I think we should stick with those definitions, and consider displaying the entire text for this specific view.

2. Can we vertically align the default preview where the beginning of the visualization matches the beginning of the floating buttons on the bottom left (do they will start on the same Y axis)

Thank you!

christianvogt commented:
@Ranelim the dynamic node width may pose a problem with the layout engine as it tries to layout nodes with an assumption of consistent node sizes.

For the first pass, we will do a fixed node width that doesn't shrink to accommodate smaller text as shown in the screenshot. We can work on refinements later on and track them in jira.

karthikjeeyar commented on src/hacbs/components/ApplicationDetails/tabs/overview/visualization/WorkflowGraph.tsx:
```diff
@@ -0,0 +1,33 @@
+  return (
+    <div className="hacbs-workflow-graph" data-test="workflow-graph">
+      <WorkflowVisualizationSurface
+        key={`${nodes.map((n) => n.id).join('-')}`}
```
Yes, key is needed here to re-layout the graph on new nodes.

karthikjeeyar commented on src/hacbs/components/ApplicationDetails/tabs/overview/visualization/hooks/useAppWorkflowData.ts:
```diff
@@ -0,0 +1,147 @@
+import { DAG } from '../../../../../topology/dag';
+import { Workflow, WorkflowNodeType } from '../types';
+
+export const useAppWorkflowData = (applicationName: string): { nodes: any[]; loaded?: boolean } => {
```
Updated.

karthikjeeyar commented:
@Ranelim 
> I believe the Node design PF is working on does not have a fixed width for all the nodes. The node width changes according to the content within it.

 Variable node length PF PR is not merged/released yet, we will refine this in later PR as christian mentioned above.
 I have created a JIRA to track https://issues.redhat.com/browse/HACBS-886.

> Can we vertically align the default preview where the beginning of the visualization matches the beginning of the floating buttons on the bottom left (do they will start on the same Y axis)

Updated it. Thanks

karthikjeeyar commented on src/hacbs/components/ApplicationDetails/tabs/overview/visualization/hooks/useAppWorkflowData.ts:
```diff
@@ -0,0 +1,150 @@
+    return contexts?.some((c) => c.name === 'application') ?? false;
+  });
+
+  const getNodeNames = (resources): string[] => resources?.map((r) => r?.metadata?.name);
```
updated. Now i am early returning if any of the resources is not loaded.

karthikjeeyar commented on src/hacbs/components/ApplicationDetails/tabs/overview/visualization/WorkflowVisualizationSurface.tsx:
```diff
@@ -0,0 +1,102 @@
+
+const WorkflowVisualizationSurface: React.FC<WorkflowVisualizationSurfaceProps> = ({ model }) => {
+  const [vis, setVis] = React.useState(null);
+  const [maxSize, setMaxSize] = React.useState(null);
```
added

karthikjeeyar commented on src/components/modal/__mocks__/tabbable.ts:
```diff
@@ -0,0 +1,11 @@
+const lib = jest.requireActual('tabbable');
+
+const tabbable = {
+  ...lib,
+  tabbable: (node, options) => lib.tabbable(node, { ...options, displayCheck: 'none' }),
+  focusable: (node, options) => lib.focusable(node, { ...options, displayCheck: 'none' }),
+  isFocusable: (node, options) => lib.isFocusable(node, { ...options, displayCheck: 'none' }),
+  isTabbable: (node, options) => lib.isTabbable(node, { ...options, displayCheck: 'none' }),
+};
+
+module.exports = tabbable;
```
with latest react-topology package, the focus-trap module is updated and it seems to have a known issue in testing environment (jsdom) [Refer](https://github.com/focus-trap/tabbable#testing-in-jsdom)

Mocking this module to avoid the following error in our Modal related tests :
```
Your focus-trap must have at least one container with at least one tabbable node in it at all times
```
cc: @christianvogt @rohitkrai03

MariaLeonova commented:
@karthikjeeyar the preview looks great. One small issue I'm seeing is the the node width: per patternfly conventions the spacer width on each side of the text should be 16 px, and currently it is more than that. That results in unnecessary amount of white space on the right, which is counter-intuitive to truncating the labels. I'm attaching a screenshot to illustrate:![Screenshot 2022-07-21 at 9 54 02](https://user-images.githubusercontent.com/14108590/180161548-de0b0aa1-cb1c-4473-ab91-e1d7cc39a8e1.png)

Can we shorten the node, so it has 16px spacing on wither side? That would also allow for more nodes to fit on a page. Thank you!

karthikjeeyar commented:
@MariaLeonova I have added equal spacing for the nodes that has truncated labels. Since the node width is fixed for the first pass, the source, build nodes will have extra spacing for now (it will be addressed as part of https://issues.redhat.com/browse/HACBS-886.)


<img width="368" alt="image" src="https://user-images.githubusercontent.com/9964343/180180944-dd494b19-c414-4fd0-915b-93caf79e5b95.png">

MariaLeonova commented:
> @MariaLeonova I have added equal spacing for the nodes that has truncated labels. Since the node width is fixed for the first pass, the source, build nodes will have extra spacing for now (it will be addressed as part of https://issues.redhat.com/browse/HACBS-886.)
> 
> <img alt="image" width="368" src="https://user-images.githubusercontent.com/9964343/180180944-dd494b19-c414-4fd0-915b-93caf79e5b95.png">

Thank you, that looks good!

karthikjeeyar commented:
/retest

christianvogt commented:
/lgtm

Ranelim commented:
Looks great!
Fixes based on the first screenshot:
Can we minimize the distance between nodes?
I'll let @MariaLeonova decide on how much but maybe as a base we can do 20px? If the arrow runs over the line then maybe 30px?

Ranelim commented:
A few more comments:

- In the [App Overview deliverable doc](https://docs.google.com/document/d/1cYK41-42c7tmurlLgTCdFgtcDrQH1WnWsMk3btY22W8/edit#), the first node specifies "Components" and not "Source code", which appeared in the marvel mocks which were outdated.
- Do we cover the visualization when a user adds a managed environment? 

Many thanks!

MariaLeonova commented:
> 

I would suggest 24 px, as it's per patternfly standards to move in 8px increments.
@Ranelim do we really need arrows for the workflow? WDYT?

karthikjeeyar commented:
@Ranelim @MariaLeonova I have updated the node distance to be 24px.

> In the [App Overview deliverable doc](https://docs.google.com/document/d/1cYK41-42c7tmurlLgTCdFgtcDrQH1WnWsMk3btY22W8/edit#), the first node specifies "Components" and not "Source code", which appeared in the marvel mocks which were outdated.

Updated

> Do we cover the visualization when a user adds a managed environment?

Yes, I have updated the screenshots that shows management environment.

MariaLeonova commented:
Hi @karthikjeeyar ! That looks good, thank you for the update! I would also suggest getting rid of the arrow, I don't think they provide any value

karthikjeeyar commented:
/retest

christianvogt commented:
/approve
/retest

karthikjeeyar commented:
E2E tests are passing locally

<img width="879" alt="image" src="https://user-images.githubusercontent.com/9964343/183099934-2af6af4f-cc2b-4230-9e22-a15f70b8dc11.png">
